pyogrio/geopandas.py

```diff
@@ -537,14 +537,19 @@ def write_dataframe(
                 geometry_type = f"{geometry_type} Z"
 
     crs = None
-    if geometry_column is not None and geometry.crs:
-        # TODO: this may need to be WKT1, due to issues
-        # if possible use EPSG codes instead
-        epsg = geometry.crs.to_epsg()
-        if epsg:
-            crs = f"EPSG:{epsg}"
-        else:
-            crs = geometry.crs.to_wkt(WktVersion.WKT1_GDAL)
+    if geometry_column is not None:
+        if geometry.crs:
+            # TODO: this may need to be WKT1, due to issues
+            # if possible use EPSG codes instead
+            epsg = geometry.crs.to_epsg()
+            if epsg:
+                crs = f"EPSG:{epsg}"
+            else:
+                crs = geometry.crs.to_wkt(WktVersion.WKT1_GDAL)
+        elif driver == "GPKG":
+            # In GPKG, None crs must be replaced by "Undefined Cartesian SRS", otherwise
+            # the default "Undefined geographic SRS" will be used.
+            crs = 'LOCAL_CS["Undefined Cartesian SRS"]'
```

Comment:
Is this enough? My understanding was that this is required:

```
ENGCRS["Undefined Cartesian SRS with unknown unit",
    EDATUM["Unknown engineering datum"],
    CS[Cartesian,2],
        AXIS["X",unspecified,
            ORDER[1],
            LENGTHUNIT["unknown",0]],
        AXIS["Y",unspecified,
            ORDER[2],
            LENGTHUNIT["unknown",0]]]
```

See the dicsussion in https://github.com/r-spatial/sf/issues/2049.

Reply:
I'm  not an expert on projections, but as far as I can tell both are equivalent: LOCAL_CS["Undefined Cartesian SRS"] is just a shorter name for the full wkt.

I did a manual check on this before, but to be sure I added an explicit check in the test that the srs_id in the GPKG written is -1, which is the main thing we are trying to do here.

Reply:
The difference is that `LOCAL_CS["Undefined Cartesian SRS"]` assumes meters as units and axes as easting and northing, which is incorrect:

```
<Engineering CRS: LOCAL_CS["Undefined Cartesian SRS",UNIT["Meter",1] ...>
Name: Undefined Cartesian SRS
Axis Info [cartesian]:
- [east]: Easting (Meter)
- [north]: Northing (Meter)
```

The WKT above is explicit about making no assumptions about anything, because we don't know anything.

Reply:
I think using that WKT does not comply to the GeoPackage specs: https://www.geopackage.org/spec/#r11. Using that WKT leads to a new custom CRS being "created" in the GeoPackage with srs_id = 10000 (or whatever is the first unused srs_id in that range in the geopackage) instead of using srs_id = -1, as should be used according to the specs.

I wonder what is the rational that they went for this solution in r-spatial? I don't really found one in the post above. Maybe it should be changed in GDAL that `ENGCRS["Undefined Cartesian SRS with unknown unit"` is (also) mapped to srs_id = -1 in geopackage?

I did some extra tests and proj does seem to treat them all as equivalent, so I suppose that in practice it probably won't matter to much with the different crs's being interused?

```
from pyproj import CRS

crs_undefined = CRS('LOCAL_CS["Undefined Cartesian SRS"]')
crs_undefined_unknown = CRS('LOCAL_CS["Undefined Cartesian SRS with unknown unit"]')
try:
    crs_undefined_unknown_eng = CRS(
        'ENGCRS["Undefined Cartesian SRS with unknown unit"]'
    )
except Exception as ex:
    print(f"Exception raised: {ex}")

crs_undefined_unknown_eng_wkt = """
    ENGCRS["Undefined Cartesian SRS with unknown unit",
        EDATUM["Unknown engineering datum"],
        CS[Cartesian,2],
            AXIS["X",unspecified,
                ORDER[1],
                LENGTHUNIT["unknown",0]],
            AXIS["Y",unspecified,
                ORDER[2],
                LENGTHUNIT["unknown",0]]]
"""
crs_undefined_unknown_eng2 = CRS(crs_undefined_unknown_eng_wkt)

# Check differences
print(f"{crs_undefined_unknown.equals(crs_undefined_unknown_eng2)=}")
print(f"{crs_undefined_unknown.to_wkt() == crs_undefined_unknown_eng2.to_wkt()=}")

print(f"{crs_undefined_unknown.equals(crs_undefined)=}")
print(f"{crs_undefined_unknown.to_wkt() == crs_undefined.to_wkt()=}")

print(f"{crs_undefined_unknown_eng2.equals(crs_undefined)=}")
print(f"{crs_undefined_unknown_eng2.to_wkt() == crs_undefined.to_wkt()=}")
```
Output:
```
Exception raised: Invalid projection: ENGCRS["Undefined Cartesian SRS with unknown unit"]: (Internal Proj Error: proj_create: Missing EDATUM / ENGINEERINGDATUM node)
crs_undefined_unknown.equals(crs_undefined_unknown_eng2)=True
crs_undefined_unknown.to_wkt() == crs_undefined_unknown_eng2.to_wkt()=False
crs_undefined_unknown.equals(crs_undefined)=True
crs_undefined_unknown.to_wkt() == crs_undefined.to_wkt()=False
crs_undefined_unknown_eng2.equals(crs_undefined)=True
crs_undefined_unknown_eng2.to_wkt() == crs_undefined.to_wkt()=False
```
```

Reply:
Tricky. When using the WKT, it correctly roundtrips with all the fields Unknown but the srid is 100000. I am not sure which is better. 

@edzer are you sure the way this is handled in {sf} is actually correct? Meaning it correctly roundtrips __and__ follows the spec?

Reply:
Reference to GDAL code treating the srs_id=-1, as found by @jorisvandenbossche :
- reading: https://github.com/OSGeo/gdal/pull/3286
- writing: https://github.com/OSGeo/gdal/pull/3312